### PR TITLE
Add move and swap axis, and vmap for slice, concat, and gather

### PR DESCRIPTION
**Pull request by awni**
## Proposed changes

Add two ops:
- `mx.moveaxis`
- `mx.swapaxes`

Add three missing vmaps:
- Concatenate
- Slice
- Gather

### CODE REVIEW

File: mlx/primitives.cpp [hunk 1]
```diff
@@ -1,8 +1,8 @@
 // Copyright © 2023 Apple Inc.
-
 #include <algorithm>
 #include <cassert>
 #include <cmath>
+#include <iostream> // TODO
```
Comment (angeloskath) [inline mlx/primitives.cpp:5]: There is a couple of forgotten `iostream` includes.

File: mlx/primitives.cpp [hunk 2]
```diff
@@ -1008,7 +1027,53 @@ std::pair<array, int> Full::vmap(
...
+    }
+    if (indices_vmapped) {
+      // Make a new index array for the vmapped dimension
+      // Reshape it so it broadcasts with other index arrays
+      // Update gather axes and slice sizes accordingly
+      auto shape = std::vector<int>(max_dims - out_ax, 1);
+      auto vmap_inds = arange(0, src.shape(out_ax), stream());
+      shape[0] = vmap_inds.shape(0);
+      vmap_inds = reshape(vmap_inds, shape, stream());
+      slice_sizes.insert(slice_sizes.begin() + out_ax, 1);
```
Comment (angeloskath) [inline mlx/primitives.cpp:1067]: I may be missing something and we should definitely merge this as is, but out of curiosity, wouldn't the above be equivalent to setting the new index to 0 and the slice size to `src.shape(out_ax)` but maybe a bit faster?

**Review by angeloskath — APPROVED** [review]
This looks fantastic and the test coverage even more so!

File: mlx/primitives.cpp (hunk 1 above)
Comment (awni) [reply mlx/primitives.cpp:5]: Just making sure you are paying attention 🤣  

sorry to keep doing this, will be more careful...

File: mlx/primitives.cpp (hunk 2 above)
Comment (awni) [reply mlx/primitives.cpp:1067]: It's exactly that when you are vmapping just over the src (see the `else` condition)

Unfortunately I think it's not that simple when you are vmapping both the indices and the `src`.

Say we have an index with shape [2, 3] and the src is shape [2, 2]. The vmap over both should have shape [2, 3].  (It's like stacking two arrays of size 3).  If we just updated the slice size, then we would get a shape [2, 3, 2].

**Pull request merged**